kolibri/core/auth/management/commands/bulkimportusers.py

```diff
@@ -0,0 +1,587 @@
+    """
+    assert len(args) > 0, "at least one argument is required"
+    if len(args) == 1:
+        # assume the first argument defines the membership
+        members = args[0].lower()
+    else:
+        members = tuple(map(str.lower, args))
+
+    def checker(value):
+        if value.lower() not in members:
+            raise ValueError(value)
+
+    return checker
+
+
+def valid_name(username=True, null=False):
+    """
+    Return a value check function which raises a ValueError if the value has
+    some of the punctuaction chars that are not allowed.
+    If username is False it allows spaces, slashes and hyphens.
+    If null is not True raises a ValueError if the supplied value is None.
+    """
+
+    def checker(v):
+        if null and v is None:
+            return checker
+        has_punc = "/[\s`~!@#$%^&*()\-+={}\[\]\|\\\/:;\"'<>,\.\?]/"  # noqa
+        if not username:
+            has_punc = "/[`~!@#$%^&*()\+={}\[\]\|\\\/:;\"'<>\.\?]/"  # noqa
+        if re.match(has_punc, v):
+            raise ValueError(v)
+
+    return checker
+
+
+class Validator(object):
+    """
+    Class to apply different validation checks on a CSV data reader.
+    """
+
+    def __init__(self, header_names):
+        self._header_names = header_names
+        self._checks = list()
+        self.classrooms = dict()
+        self.coach_classrooms = dict()
+        self.users = dict()
+        self.roles = {r: list() for r in roles_map.values() if r is not None}
+
+    def add_check(self, header_name, check, message):
+        """
+        Add a header check, i.e., check whether the header record is consistent
+        with the expected field names.
+        `header_name` - name of the header for the column to be checked
+        `check`- function to be used as validator of the values in the column
+        `message` - problem message to report if a value is not valid
+        """
+        self._checks.append((header_name, check, message))
+
+    def get_username(self, row):
+        username = row.get("Username")
+        if username in self.users.keys():
+            return None
+
+        return username
+
+    def check_classroom(self, row, username):
+        def append_users(group, key):
+            try:
+                groups = row.get(key, None).split(",")
+                for classroom in groups:
+                    if not classroom:
+                        continue
+                    if classroom in group:
+                        group[classroom].append(username)
+                    else:
+                        group[classroom] = [
+                            username,
+                        ]
+            except AttributeError:
+                # there are not members of 'key'
+                pass
+
+        # enrolled learners:
+        append_users(self.classrooms, "Enrolled in")
+
+        # assigned coaches
+        user_role = row.get("User type", "learner").lower()
```

Comment:
We shouldn't assume that the headers will be in English. Perhaps we can instead make constants which refer to column numbers, e.g.

```python
USER_TYPE_COL = 3
```

etc...?

Reply:
As Ivan mentioned in a thread, I think translating headers is quite dangerous. It's just a data type. I still remember how Microsoft regreted after translating VB for applications in the first Excel versions and had to roll it back later... I am not sure it's a good idea, because it will be really hard to validate the headers names, and to map them to our right field names.
In any case, if we decide to do it, I think this is something similar to the messages errors question: in case we do it it can be done later, after the code is tested and works correctly.

Reply:
I think translating headers is very different from translating enumerated constants because columns can be determined by column number independently of their name

Reply:
I guess the question relates to our ability to extend the import format later:

1. Do we add more columns but are forced to keep the order?
2. Can we add and remove columns as long as we keep the same names?

(1) is easier for i18n but (2) is better for long-term flexibility of the format